core/src/main/java/org/infinispan/stream/impl/LocalStreamManagerImpl.java

```diff
@@ -298,41 +309,51 @@ private void handleResponseError(CompletionStage<ValidResponse> rpcFuture, Objec
                  segments);
       }
       CacheSet<Original> originalSet = toOriginalSet(includeLoader, entryStream);
-      SegmentListener listener = new SegmentListener(segments, operation);
       R value;
 
       operation.handleInjection(registry);
-      // We currently only allow 1 request per id (we may change this later)
-      changeListener.put(requestId, listener);
-      if (trace) {
-         log.tracef("Registered change listener for %s", requestId);
+      SegmentListener listener;
+      if (requiresListener) {
+         listener = new SegmentListener(segments, operation);
+         // We currently only allow 1 request per id (we may change this later)
+         changeListener.put(requestId, listener);
+         if (trace) {
+            log.tracef("Registered change listener for %s", requestId);
+         }
+      } else {
+         listener = null;
       }
+      IntSet lostSegments;
       try {
          operation.setSupplier(() -> getRehashStream(originalSet, requestId, listener, parallelStream, entryStream,
                segments, keysToInclude, keysToExclude));
          value = operation.performOperation();
+         if (listener != null) {
+            lostSegments = listener.segmentsLost;
+         } else {
+            lostSegments = IntSets.immutableEmptySet();
+         }
          if (trace) {
             log.tracef("Request %s completed for segments %s with %s suspected segments", requestId, segments,
-                    listener.segmentsLost);
+                    lostSegments);
          }
       } finally {
-         changeListener.remove(requestId);
-         if (trace) {
-            log.tracef("UnRegistered change listener for %s", requestId);
+         if (listener != null) {
+            changeListener.remove(requestId);
+            if (trace) {
+               log.tracef("UnRegistered change listener for %s", requestId);
+            }
          }
       }
       if (cache.getStatus() != ComponentStatus.RUNNING && cache.getStatus() != ComponentStatus.INITIALIZING) {
          if (trace) {
             log.tracef("Cache status is no longer running, all segments are now suspect for %s", requestId);
          }
-         listener.segmentsLost.addAll(segments);
+         lostSegments = segments;
```

Comment:
IMO we shouldn't send a `StreamResponseCommand` at all if the cache is shutting down.

Reply:
Hrmm, I will look into this. As long as we don't send a response to the command, this should be okay as I would expect to get a `SuspectException` at https://github.com/infinispan/infinispan/blob/master/core/src/main/java/org/infinispan/stream/impl/ClusterStreamManagerImpl.java#L210 Or maybe it is `CacheNotFoundException`? Either way I can look into it if it all fails.

Reply:
`SingleResponseCollector.validOnly()` throws a `SuspectException`. `CacheNotFoundException` is a server thing :)

Reply:
Alright :) Actually give me a couple more hours to look at this again, I want to verify something.

Reply:
After sleeping on it I realized we can't just remove the response. This will cause a timeout for the caller, since the response is completed based on either it sending the `StreamResponseCommand` or the `StreamRequestCommand` returning a `SuspectException`. If we just return normally it won't get a `SuspectException` so we have to send this command for now. We could revisit changing this, but it requires more changes than I would like to introduce this late in the cycle.

Reply:
Have to look a little further, doesn't seem to be the case.

Reply:
Yeah so it seems to be a timing issue if I take it out. Sometimes it gets the `SuspectException` and sometimes not. I added in some trace messages to see when it happens in a test:
```
09:55:19,454 FATAL (remote-thread-DistributedStreamRehashStressTest-NodeAO-p322-t11) [LocalStreamManagerImpl] Cache status is no longer running, all segments are now suspect for DistributedStreamRehashStressTest-NodeA-43398#371
09:55:29,080 FATAL (remote-thread-DistributedStreamRehashStressTest-NodeAT-p362-t10) [LocalStreamManagerImpl] Cache status is no longer running, all segments are now suspect for DistributedStreamRehashStressTest-NodeA-43398#422
09:55:39,083 FATAL (ForkThread-1,DistributedStreamRehashStressTest) [DistributedCacheStream] Timeout encountered with id: DistributedStreamRehashStressTest-NodeA-43398#422
```

So as you can see the first time was okay, more than likely because it got the `SuspectException` back before it could return the empty response.

Reply:
This needs me to change it to send a response instead of a `StreamResponseCommand` which is a bit much for this PR imo.